### PR TITLE
e2e: tests for TCPRoute

### LINKED ISSUE #2010: Add E2E tests for TCPRoute
*Description*:
>Describe the issue.

Add E2E tests in https://github.com/envoyproxy/gateway/tree/main/test/e2e/tests to test [TCPRoute](https://gateway-api.sigs.k8s.io/guides/tcp/)

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

### PR DESCRIPTION
**Pull request by slayer321**
**What type of PR is this?**

e2e: tests for TCPRoute

**What this PR does / why we need it**:
Add the e2e test for TCPRoute
**Which issue(s) this PR fixes**:

Fixes #2010

### CODE REVIEW

File: test/e2e/tests/tcp-route.go [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+					Path: "/",
+				},
+				Response: http.Response{
+					StatusCode: 200,
+				},
+				Namespace: ns,
+			}
+
+			// Send a request to an valid path and expect a successful response
+			http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, OkResp)
```
Comment (arkodg) [inline test/e2e/tests/tcp-route.go:82]: how is the destination port (8090, 8080) being computed ?

Comment (slayer321) [reply test/e2e/tests/tcp-route.go:82]: As I understand following the [docs](https://gateway-api.sigs.k8s.io/guides/tcp/).  We have one Gateway resource(and two TCP listerns on port 8080 and 8090) and two TCPRoute resources. When e2e test is run gateways are created and requests are passed to the added services port for TCPRoute.  (Please correct me if I'm missing something here.)

Comment (slayer321) [reply test/e2e/tests/tcp-route.go:82]: While checking for `WaitForGatewayAddress` function found that it only get's the [port of first listeners](https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/kubernetes/helpers.go#L411) . In our case we have two listeners sets on (8080, 8090) , but as It is just getting the first port of the listener i.e `8080` . After running the e2e test I found that it is making request on same port for both listeners

```
tcp-route.go:83: Making GET request to http://172.18.255.205:8080/
tcp-route.go:102: Making GET request to http://172.18.255.205:8080/
```

Comment (arkodg) [reply test/e2e/tests/tcp-route.go:82]: can you ensure traffic is sent to different ports

Comment (slayer321) [reply test/e2e/tests/tcp-route.go:60]: Done, Now traffic is send to specific ports
```
tcp-route.go:78: Making GET request to http://172.18.255.205:8090/
tcp-route.go:60: Making GET request to http://172.18.255.205:8080/
```

File: test/e2e/tests/tcp-route.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+		TCPRouteMustHaveParents(t, c, timeoutConfig, routeNN, parents, namespaceRequired)
+	}
+
+	return gwAddr
+
+}
+
+// WaitForGatewayAddress waits until at least one IP Address has been set in the
+// status of the specified Gateway.
+func WaitForGatewayAddress(t *testing.T, client client.Client, timeoutConfig config.TimeoutConfig, gwName types.NamespacedName, sectionName string) (string, error) {
```
Comment (arkodg) [inline test/e2e/tests/tcp-route.go:132]: can't this be reused 
https://github.com/kubernetes-sigs/gateway-api/blob/95a6b8a1f7b27bc0e632e4f44fc3e277a80276bc/conformance/utils/kubernetes/helpers.go#L394 ?

can the other helper functions be reused from upstream ?

**Review by arkodg — APPROVED** [review]
LGTM thanks

**Review by zirain — APPROVED** [review]
Thanks, hope we can move these to upstream one day.

**Pull request merged**